docs/rfc-fluffy-unicorn.md

```diff
@@ -0,0 +1,152 @@
+__First things first:__ Fluffy-Unicorn is based on the work of a number of projects that have been adapted to work together and fit the requirements of nimona. [JSON], [JSON-LD], [JSON-Schema], [Tagged JSON], and [Object hash].
+
+## Types
+Fluffy-Unicorn has its own list of basic types and structures, more closely resembling the ones of [JSON], other than the split of number to int/uint/float, and the addition of a binary type.
+
+* Object - A collection of key/value pairs
+* Array - An ordered list of values
+* String
+* Integer
+* Unsigned Integer
+* Float
+* Data (binary data)
+* Boolean
+
+Continuing its comparison with JSON, there are some restrictions that are not present in the JSON format.
+
+* The top-level Fluffy-Unicorn-Object must always be an object
+* All values of an array must be of the same type.
+* A number of key names are reserved.
+
+## Type Hinting
+In order to allow Fluffy-Unicorn-Objects to be encoding and language agnostic, the types of all values have to be able to survive multiple re-encodings between incompatible formats, languages, and libraries.
+
+Since we can’t depend on each encoding format to keep the correct type, we provide a hint for the value’s type as part of the key. 
+
+```json
+{
+  "some-string:s": "bar",
+  "nested-object:o": {
+    "unsigned-number-one:u": 1,
+    "array-of-ints:a<i>": [-1, 0, 1]
+  }
+}
+```
+
+Hints for all of the supported types and structures:
+
+* `s` string
+* `i` int
+* `u` uint
+* `f` float
+* `d` data (binary)
+* `b` bool
+* `a` array
+  * `a<x>` array of type `x`
+	  * `a<a<x>>` nested array of type `x`
+* `o` object (map, dict)
+* `o*` object reference _We’ll come back to this one later_
+
+_Note: Type hinting is based on the [Tagged JSON] micro-format with a couple of minor changes; mainly terminology, the hints/tags used, the removal of sets and binary data formats, and the addition of the object reference type._
+
+## Self-description & Schemas
+Fluffy-Unicorn takes advantage of the reserved key `$schema` to allow objects to define their schemas. Defining a schema is not required.
+
+The value of `$schema` is an object that defines and documents the structure of the object.
+
+```json
+{
+  "$schema:o": {
+    "description:s": "Hello world object",
+    "properties:a<s>": [
+      "body:s"
+    ]
+  },
+  "body:s": "Hello world!"
+}
+```
+
+_Note: The schema definition is still work in progress and will remain like this for a while._
+
+## Object Hashing
+Fluffy-Unicorn-Object-Hashing is based off the JSON version of [Ben Laurie]’s [Object hash] and provides a way to consistently hash objects that works across languages and encodings.
+
+Given an object, it will sort its key/value pairs based on their key, loop through them, hash each key, and then using the type hint in the key, normalise and hash its value (after prefixing them with the type’s hint). 
+It will finally gather all key and value hashes, and hash them all together to produce the object’s hash.
```

Comment:
What's the advantage of using kv hashes to make the object hash instead of using the nornalized merged kv strings?

Reply:
This is what enables the value redaction.

If you want to keep a value hidden from others, instead of including the actual value in the object you send them, you simply send the value's hash in its place. Now when the hashing function loops through the key/values, it will see that this is already a hash, and simply use it instead of hashing it.

This allows removing data from the object without affecting the hash.